src/libstd/keyword_docs.rs

```diff
@@ -1578,12 +1578,255 @@ mod type_keyword {}
+/// }
+/// ```
+///
+/// Traits can also be declared as `unsafe`:
+///
+/// ```rust
+/// unsafe trait UnsafeTrait {}
+/// ```
+///
+/// Documenting public `unsafe fn` (and other elements) is important. The
+/// standard library has many examples of this, like the following which is an
+/// extract from [`Vec::set_len`]. The `# Safety` section explains the contract
+/// that must be fulfilled to safely call the function. If necessary, a
+/// `# Panics` section can be added to document possible panics and their
+/// reasons.
+///
+/// ```rust,ignore (stub-to-show-doc-example)
+/// /// Forces the length of the vector to `new_len`.
+/// ///
+/// /// This is a low-level operation that maintains none of the normal
+/// /// invariants of the type. Normally changing the length of a vector
+/// /// is done using one of the safe operations instead, such as
+/// /// `truncate`, `resize`, `extend`, or `clear`.
+/// ///
+/// /// # Safety
+/// ///
+/// /// - `new_len` must be less than or equal to `capacity()`.
+/// /// - The elements at `old_len..new_len` must be initialized.
+/// pub unsafe fn set_len(&mut self, new_len: usize)
+/// ```
+///
+/// ## Using `unsafe` blocks and `impl`s
+///
+/// Calling an `unsafe` function requires an `unsafe {}` block, which can take
+/// two forms. Note that the `enclosing_fn` form is the subject of [proposal]
+/// to modify its behavior as an `unsafe {}` block.
+///
+/// ```rust
+/// # #![allow(dead_code)]
+/// unsafe fn unsafe_fn() {}
+///
+/// // This function is unsafe to call and unsafe functions can be called
+/// // inside without problems.
+/// unsafe fn enclosing_unsafe() {
+///     unsafe_fn()
+/// }
 ///
+/// // This function is safe to call. Any call to an unsafe function inside
+/// // must be enclosed in an unsafe block. `interior_unsafe` is responsible
+/// // for ensuring that memory unsafety never occurs, no matter how or when
+/// // it is called.
+/// fn interior_unsafe() {
+///     // SAFETY: `unsafe_fn` does nothing.
+///     unsafe { unsafe_fn() }
+/// }
+/// ```
+///
+/// Traits marked as `unsafe` must be [`impl`]emented using `unsafe`. This makes
+/// a guarantee to other `unsafe` code that a property holds for this type. It
+/// is undefined behavior to make an `unsafe impl` for a type which does not
+/// have this property. The [Send] and [Sync] are examples of traits that are
+/// `unsafe` to implement.
+///
+/// ```rust
+/// # #![allow(dead_code)]
+/// unsafe trait UnsafeTrait {}
+///
+/// struct S;
+///
+/// // SAFETY: explanation about why this impl is safe.
+/// unsafe impl UnsafeTrait for S {}
+/// ```
+///
+/// ## Methods inside `unsafe trait`
+///
+/// Methods inside an `unsafe trait` are **not** inherently `unsafe`:
+///
+/// ```rust
+/// /// Implementors of this trait must guarantee `even` and `even_unsafe`
+/// /// always return an even number.
+/// unsafe trait EvenGetter {
+///     fn even() -> i32;
+///     unsafe fn even_unsafe() -> i32;
```

Comment:
What is the safety contract of this function? You said yourself that this is important, now your own example violates your guidelines. ;)

Reply:
Or this could be `even_unchecked` which does no validation, faster. Maybe `even` might not be a good example.

Reply:
Oups, forgot that 😅 

I changed the example, I hope it is better now.